lib/common_test/doc/src/ct_suite.xml

```diff
@@ -0,0 +1,619 @@
+    </copyright>
+    <legalnotice>
+      Licensed under the Apache License, Version 2.0 (the "License");
+      you may not use this file except in compliance with the License.
+      You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+      Unless required by applicable law or agreed to in writing, software
+      distributed under the License is distributed on an "AS IS" BASIS,
+      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+      See the License for the specific language governing permissions and
+      limitations under the License.
+
+      The Initial Developer of the Original Code is Ericsson AB.
+    </legalnotice>
+    <title>ct_suite</title>
+    <prepared></prepared>
+    <docno></docno>
+    <date></date>
+    <rev></rev>
+  </header>
+  <module since="">ct_suite</module>
+  <modulesummary>-behaviour(ct_suite).
+  </modulesummary>
+  <description>
+    <p>The following section describes the mandatory and optional test suite
+      functions that <c>Common Test</c> calls during test execution.
+      For more details, see section
+      <seeguide marker="write_test_chapter">Writing Test Suites</seeguide>
+      in the User's Guide.</p>
+  </description>
+
+  <datatypes>
+    <datatype>
+      <name name="ct_testname" n_vars="0"/>
+      <desc><p>The name of the testcase function.</p></desc>
+    </datatype>
+    <datatype>
+      <name name="ct_groupname" n_vars="0"/>
+      <desc><p>The name of the test group.</p></desc>
+    </datatype>
+    <datatype>
+      <name name="ct_config" n_vars="0"/>
+      <desc><p>The configuration data that can be modified.</p></desc>
+    </datatype>
+    <datatype>
+      <name name="ct_status" n_vars="0"/>
+      <desc><p>The status value for a nested subgroup.</p></desc>
+    </datatype>
+    <datatype>
+      <name>ct_group_def()</name>
+      <desc><p>The test group definition, as returned by <seemfa marker="#Module:groups/0"><c>Module:groups/0</c></seemfa>.</p></desc>
+    </datatype>
+    <datatype>
+      <name>ct_test_def()</name>
+      <desc><p>The test suite definition, as returned by <seemfa marker="#Module:all/0"><c>Module:all/0</c></seemfa>.</p></desc>
+    </datatype>
+    <datatype>
+      <name>ct_info()</name>
+      <desc><p>The test suite information, as returned by <seemfa marker="#Module:suite/0"><c>Module:suite/0</c></seemfa>, <seemfa marker="#Module:group/1"><c>Module:group/1</c></seemfa> and <seemfa marker="#Module:Testcase/0"><c>Module:Testcase/0</c></seemfa>.</p></desc>
+    </datatype>
+  </datatypes>
+
+  <funcs>
+    <fsdescription>
+      <title>Callback Functions</title>
+      <p>
+        The following functions are to be exported from a
+        <c>ct_suite</c> callback module in order to define
+        the callback interface for a test suite.
+      </p>
+    </fsdescription>
+
+    <func>
+      <name since="">Module:all() -> [ct_test_def()] | {skip, Reason}</name>
```

Comment:
All type information can be auto-generated from the specs by following the pattern below. This is a general comment for all function definitions.
 
```xml
    <func>
      <name name="all" arity="0" since=""/>
      <fsummary> ... 1-liner summary...</fsummary>
      <desc>
	<p> ... details ...</p>
      </desc>
    </func>
```

Reply:
I don't understand your comment (what is "to follow a pattern"?). Is there something you wish me to change? Looking at the generated documentation it seems good to me.

Reply:
You mean removing `<type>`?
I can do that, but will that work with what @IngelaAndin and I discussed before [here](https://github.com/erlang/otp/pull/2794#issuecomment-713422455)?

Reply:
I'm now getting `Error in ct_suite:all/0: unknown spec` for
```
      <name name="all" arity="0" since=""/>
```

Reply:
The "pattern" I meant was to try to auto-generate the type sections using the specs in the code. I did a fast prototype to see if it works and I realized that the doc generation is missing some functionality when documenting callback functions. There are ways to work around the limitations but it is not that elegant. Try the diff below to see the difference. Ideally the types section should be generated from the specs with working links. I'll discuss this with team and I'll come back soon.

```diff
diff --git a/lib/common_test/doc/src/ct_suite.xml b/lib/common_test/doc/src/ct_suite.xml
index 0289a569f0..eb8dea7735 100644
--- a/lib/common_test/doc/src/ct_suite.xml
+++ b/lib/common_test/doc/src/ct_suite.xml
@@ -61,8 +61,8 @@
       <desc><p>The test group definition, as returned by <seemfa marker="#Module:groups/0"><c>Module:groups/0</c></seemfa>.</p></desc>
     </datatype>
     <datatype>
-      <name>ct_test_def()</name>
-      <desc><p>The test suite definition, as returned by <seemfa marker="#Module:all/0"><c>Module:all/0</c></seemfa>.</p></desc>
+      <name name="ct_test_def"/>
+      <desc><p>The test suite definition, as returned by <seemfa marker="#all/0"><c>Module:all/0</c></seemfa>.</p></desc>
     </datatype>
     <datatype>
       <name>ct_info()</name>
@@ -81,22 +81,9 @@
     </fsdescription>
 
     <func>
-      <name since="">Module:all() -> [ct_test_def()] | {skip, Reason}</name>
+      <name name="all" arity="0" since=""/>
       <fsummary>Returns the list of all test case groups and test cases
 	in the module.</fsummary>
-      <type>
-	<v><seetype marker="#ct_test_def">ct_test_def()</seetype> = TestCase | {group, GroupName} | {group, GroupName, Properties} | {group, GroupName, Properties, SubGroups}</v>
-	<v>TestCase = <seetype marker="#ct_testname">ct_testname()</seetype></v>
-	<v>GroupName = <seetype marker="#ct_groupname">ct_groupname()</seetype></v>
-	<v>Properties = [parallel | sequence | Shuffle | {RepeatType, N}] | default</v>
-	<v>SubGroups = [{GroupName, Properties} | {GroupName, Properties, SubGroups}]</v>
-	<v>Shuffle = shuffle | {shuffle, Seed}</v>
-	<v>Seed = {integer(), integer(), integer()}</v>
-	<v>RepeatType = repeat | repeat_until_all_ok | repeat_until_all_fail | repeat_until_any_ok | repeat_until_any_fail</v>
-	<v>N = integer() | forever</v>
-	<v>Reason = term()</v>
-      </type>
-
       <desc>
 	<p>MANDATORY</p>
 
diff --git a/lib/common_test/src/ct_suite.erl b/lib/common_test/src/ct_suite.erl
index a2d23e15ef..6f3c287593 100644
--- a/lib/common_test/src/ct_suite.erl
+++ b/lib/common_test/src/ct_suite.erl
@@ -12,6 +12,8 @@
               ct_info/0
              ]).
 
+-export([all/0]).
+
 -type ct_testname() :: atom().
 -type ct_groupname() :: atom().
 -type ct_config() :: [{Key :: atom(), Value :: term()}].
@@ -77,6 +79,8 @@
             ].
 -type ct_test_def() :: ct_testname() | ct_group_ref() | ct_testcase_ref().
 
+
+
 -callback all() ->
     [TestDef :: ct_test_def()] |
     {skip, Reason :: term()}.
@@ -128,3 +132,10 @@
                      init_per_testcase/2,
                      end_per_testcase/2
                     ]).
+
+-spec all() -> [TestDef] | {skip, Reason} when
+      TestDef :: ct_test_def(),
+      Reason :: term().
+
+all() ->
+    [].

```

Reply:
I'll wait for your feedback before moving forward. Thanks.

Reply:
We have decided to handle this in a future improvement.